src/alert_worker.rs

```diff
@@ -40,119 +58,128 @@ pub async fn alert_worker(
     let mut con = client_redis
         .get_multiplexed_async_connection()
         .await
-        .unwrap();
+        .map_err(AlertWorkerError::ConnectRedisError)?;
     let queue_name = format!("{}_alerts_packets_queue", stream_name);
     let queue_temp_name = format!("{}_alerts_packets_queuetemp", stream_name);
     let classifer_queue_name = format!("{}_alerts_classifier_queue", stream_name);
 
-    let mut alert_counter = 0;
     let command_interval = worker_util::get_check_command_interval(config_file, &stream_name);
+    let mut command_check_countdown = command_interval;
     // ALERT SCHEMA (for fast avro decoding)
-    let schema = ztf_alert_schema().unwrap();
+    let schema = ztf_alert_schema().ok_or(AlertWorkerError::GetAlertSchemaError)?;
     let mut count = 0;
     let start = std::time::Instant::now();
     loop {
         // check for command from threadpool
-        if alert_counter - command_interval > 0 {
-            alert_counter = 0;
-            if let Ok(command) = receiver.lock().unwrap().try_recv() {
-                match command {
-                    WorkerCmd::TERM => {
-                        warn!("alert worker {} received termination command", id);
-                        return;
-                    }
+        if command_check_countdown == 0 {
+            match receiver.try_recv() {
+                Ok(WorkerCmd::TERM) => {
+                    info!("alert worker {} received termination command", id);
+                    break;
+                }
+                Err(TryRecvError::Disconnected) => {
+                    warn!("alert worker {} receiver disconnected, terminating", id);
+                    break;
+                }
+                Err(TryRecvError::Empty) => {
+                    command_check_countdown = command_interval;
                 }
             }
         }
         // retrieve candids from redis
-        let result: Option<Vec<Vec<u8>>> =
-            con.rpoplpush(&queue_name, &queue_temp_name).await.unwrap();
+        let Some(mut value): Option<Vec<Vec<u8>>> = con
+            // TODO: If the intention is to have a reliable queue, then there
+            // needs to be a separate thread that continuously monitors the temp
+            // queue and adds any items it finds back to the main queue.
+            // However, we're already pushing the value back onto the when we
+            // handle a processing error, so do we need the temp queue?
+            .rpoplpush(&queue_name, &queue_temp_name)
+            .await
+            .map_err(AlertWorkerError::PopAlertError)?
+        else {
+            info!("ALERT WORKER {}: Queue is empty", id);
+            tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
+            command_check_countdown = 0;
+            continue;
+        };
+        // TODO: Where does the nested vector structure come from, and are there
+        // any guarantees about whether the outer vector always has exactly one
```

Comment:
yes, we receive alerts one by one and the avro will always have exactly one record.

Reply:
Oh that said, the kafka consumer should push only a Vec<u8> not a nested Vec, I think we were forced to do this for some reason in the kafka consumer, while encountering an error. I'm sure we can change it

Reply:
Yeah I dig into the consumer code, too and it looked to me like it's pushing `Vec<u8>`. I don't understand how we're getting a `Vec<Vec<u8>>` out of the queue.

Reply:
I think it's just because when we read from it we read a batch of arbitrary size (which here we force to 1), so that's why it's wrapped in another Vec.